modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/objectbuilder/DefaultObjectBuilder.java

```diff
@@ -114,10 +120,34 @@ public boolean isDynamic() {
   public T build(ValueResolvingContext context) throws MuleException {
     T object = createInstance(prototypeClass);
 
+    final CursorComponentDecoratorFactory componentDecoratorFactory =
+        context != null
+            ? (CursorComponentDecoratorFactory) context.getProperty(COMPONENT_DECORATOR_FACTORY_KEY)
+            : null;
+
     for (Map.Entry<FieldSetter, ValueResolver<Object>> entry : resolvers.entrySet()) {
+      final Object resolvedValue = resolveValue(entry.getValue(), context);
       entry.getKey().set(object,
-                         context == null || context.resolveCursors() ? resolveCursor(resolveValue(entry.getValue(), context))
-                             : resolveValue(entry.getValue(), context));
+                         context == null || context.resolveCursors()
+                             ? resolveCursor(resolvedValue,
```

Comment:
Maybe in the future we should add a kind of visitor of a class that uses double dispatch so we can unify the uses of the decoration here and in ParameterGroupBuilder and possibly in the case of the output decorators. Two many ifs.